reference/MarkdownViewer/main.cpp

```diff
@@ -0,0 +1,73 @@
+#include <CtrlLib/CtrlLib.h>
+
+#include <plugin/md/Markdown.h>
+#include <plugin/bmp/bmp.h>
+#include <plugin/jpg/jpg.h>
+
+using namespace Upp;
+
+struct MarkdownViewer : TopWindow {
+	typedef MarkdownViewer CLASSNAME;
+	RichTextView view;
+	MenuBar menubar;
+	Array<QtfRichObject> images;
+	
+	MarkdownViewer()
+	{
+		Title(t_("A simple markdown viewer"));
+		Sizeable().Zoomable().SetRect(0, 0, 800, 600);
+		Add(view.SizePos());
+		AddFrame(menubar);
+		menubar.Set(THISFN(MainMenu));
+	}
+	
+	void MainMenu(Bar& menu)
+	{
+		menu.Sub(t_("File"), THISFN(FileMenu));
+	}
+	
+	void FileMenu(Bar& menu)
+	{
+		menu.Add(t_("Open"), THISFN(OpenFile)).Key(K_CTRL_O);
+		menu.Add(!IsNull(~view), t_("Copy to clipboard (as qtf)"), THISFN(CopyAsQtf)).Key(K_CTRL_C);
+		menu.Separator();
+		menu.Add(t_("Exit"),   THISFN(Close)).Key(K_CTRL_Q);
+	}
+	
+	void OpenFile()
+	{
+		if(String f = SelectFileOpen("*.md"); !f.IsEmpty())
+			ParseFile(f);
+	}
+	
+	void ParseFile(const String& fname)
+	{
+		images.Clear();
+		MarkdownConverter mdc;
+		mdc.WhenImages = [&](VectorMap<String, String>& imglist)
+		{
+			String dir = GetFileDirectory(fname);
+			Progress pi(t_("Loading images..."), imglist.GetCount());
+			pi.pi.Percent();
+			for(int i = 0; i < imglist.GetCount(); i++) {

```

Comment:
For each loop can be used instead.

Reply:
Is this necessary? I really don't really like to bring in a std::library function when it is not more useful than the traditional for loop. IMO, it doesn't really add anything here.